core/services/pg/connection.go

```diff
@@ -13,11 +13,11 @@ import (
 	"github.com/smartcontractkit/chainlink/v2/core/store/dialects"
 )
 
-type ConnectionConfig interface {
-	DatabaseDefaultIdleInTxSessionTimeout() time.Duration
-	DatabaseDefaultLockTimeout() time.Duration
-	MaxOpenConns() int
-	MaxIdleConns() int
+type ConnectionConfig struct {
+	DefaultIdleInTxSessionTimeout time.Duration
+	DefaultLockTimeout            time.Duration
+	MaxOpenConns                  int
+	MaxIdleConns                  int
```

Comment:
I'm not opposed to switching over to structs, but what is the rule for when to do this vs. leave it as an interface?

Reply:
One tradeoff that I'm not sure we considered is that these don't nest for extension as nicely as interfaces - the embedding doesn't always work as expected.

Reply:
That's a good question! I've been moving to structs whenever all of the fields that I'm touching have been ported over to the new interface.

So in the example above, I moved this over to a struct since I moved over the last two fields.

In the case of LockedDBConfig I left it as an interface since we've yet to move over DatabaseURL and one other field if I remember correctly.

Reply:
@jmank88 Agreed! That and the verbosity is a downside to this approach. In the case of an overlapping set, I would argue that we should just do an explicit conversion between the two configs rather than using embedding. What do you think of that as a workaround?

Reply:
Don't we still pay the cost of the extra type definitions and field assignments though? That's what I was worried about.

Reply:
We also give up the implicit nature of interfaces - might not matter but could sneak up on us later :thinking:

Reply:
> But I think that's unavoidable in this case unless we're happy to accept larger interfaces in our components 

I'm not following this. The advantage of interfaces over structs is that we _can_ use smaller sets of fields and implicitly pass along something larger without rigging a bunch of extra boilerplate.

> The problem we have at the moment is that the Config interface behaves a bit like a God object, and because we're implicitly taking subsets of that interface everywhere the core interface therefore becomes very hard to change. (That's why this refactor is so hard.)

Moving from one config interface to a hierarchy is orthogonal to choosing between structs vs. interfaces for the leaves of the tree.

> If we want to make it easier to change, we have to add abstraction points, i.e. new types so that we can convert between them.

What do you mean? The extra boilerplate makes changes less practical for a developer, because they can no longer refactor a common shared method across everything at once like we can do with interfaces.